### PR TITLE
VizPanel: Support panel migrations and state changes 

### PR DESCRIPTION
**Pull request by torkelo**
TODO
* [x] Add unit tests
* [x] Handle migrations 
* [x] Handle widget style panels that don't need data better 
* [x] When creating VizPanel from old PanelModel set pluginVersion (needed for correct migrations) 

Now dashboards like the bar gauge dashboards works:
http://localhost:3000/scenes/dashboard/xMsQdBfWz 

http://localhost:3000/scenes/dashboard/vmie2cmWz22

### CODE REVIEW

**Comment by mckn** [comment]
Awesome work! Regarding migrations, have you thought anything about this long term? I guess we will use the schema pipeline to do the migration before returning the options to the client?

Lots of questions here and I'm not expecting any answers. Just curious if you have given this any thought?

File: public/app/features/scenes/core/SceneTimeRange.tsx [hunk 1]
```diff
@@ -1,9 +1,17 @@
-import { TimeRange, UrlQueryMap } from '@grafana/data';
+import { getDefaultTimeRange, getTimeZone, TimeRange, UrlQueryMap } from '@grafana/data';
 
 import { SceneObjectBase } from './SceneObjectBase';
 import { SceneObjectWithUrlSync, SceneTimeRangeState } from './types';
 
 export class SceneTimeRange extends SceneObjectBase<SceneTimeRangeState> implements SceneObjectWithUrlSync {
+  public constructor(state: Partial<SceneTimeRangeState>) {
```
Comment (ivanortegaalba) [inline public/app/features/scenes/core/SceneTimeRange.tsx:7]: I don't like the need to set an empty object to create a `SceneTimeRange`. Why don't we allow creating a scene without arguments?
```suggestion
  public constructor(state: Partial<SceneTimeRangeState> = {}) {
```

Comment (torkelo) [reply public/app/features/scenes/core/SceneTimeRange.tsx:7]: @ivanortegaalba yea, agree. That is a good idea

File: public/app/features/scenes/components/VizPanel/VizPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    const sceneTimeRange = sceneGraph.getTimeRange(this);
+    sceneTimeRange.setState({
+      raw: {
+        from: toUtc(timeRange.from),
+        to: toUtc(timeRange.to),
+      },
+      from: toUtc(timeRange.from),
+      to: toUtc(timeRange.to),
+    });
+  };
```
Comment (ivanortegaalba) [inline public/app/features/scenes/components/VizPanel/VizPanel.tsx:106]: Is it on purpose to not pass the timezone? I've been testing around and I saw the timezone is not applied when switching between scenes and dashboards.

File: public/app/features/scenes/components/VizPanel/VizPanelRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        leftItems={isDraggable ? [dragHandle] : undefined}
+      >
+        {(innerWidth, innerHeight) => (
+          <>
+            {!dataWithOverrides && <div>No data...</div>}
+            {dataWithOverrides && (
+              <ErrorBoundaryAlert dependencies={[plugin, data]}>
+                <PluginContextProvider meta={plugin.meta}>
+                  <PanelComponent
+                    id={1}
```
Comment (ivanortegaalba) [inline public/app/features/scenes/components/VizPanel/VizPanelRenderer.tsx:67]: Out of curiosity. Why are we fixing the `id` to a constant here?

File: public/app/features/scenes/querying/SceneQueryRunner.test.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+      await new Promise((r) => setTimeout(r, 1));
+
+      expect(queryRunner.state.data?.state).toBe(LoadingState.Done);
+      // Default max data points
+      expect(sentRequest?.maxDataPoints).toBe(500);
+    });
+  });
+
+  describe('when activated and waitForWidth set to true', () => {
```
Comment (ivanortegaalba) [inline public/app/features/scenes/querying/SceneQueryRunner.test.ts:58]:
```suggestion
  describe('when activated and maxDataPointsFromWidth set to true', () => {
```
Small tweak, but it makes me go to the source file to see if there is any boolean called waitForWidth

File: public/app/features/scenes/querying/SceneQueryRunner.test.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+      expect(queryRunner.state.data).toBeUndefined();
+
+      queryRunner.activate();
+
+      await new Promise((r) => setTimeout(r, 1));
+
+      expect(queryRunner.state.data?.state).toBeUndefined();
+
+      queryRunner.setContainerWidth(1000);
```
Comment (ivanortegaalba) [inline public/app/features/scenes/querying/SceneQueryRunner.test.ts:74]: After this, and before going to the next cycle, what's the LoadingState of the data? Loading or undefined?

**Review by ivanortegaalba — APPROVED** [review]
All looks pretty good to me! I added some questions, probably because I miss some details :) 

Also, I was playing with different visualization, and the one breaking the most was [the table](http://localhost:3000/d/1YGJxPUnz/table-panel-pagination?orgId=1). I don't know if this is something to worry about or if it is because we haven't done all migrations and wiring yet.

File: public/app/features/scenes/components/VizPanel/VizPanelRenderer.tsx (hunk 3 above)
Comment (torkelo) [reply public/app/features/scenes/components/VizPanel/VizPanelRenderer.tsx:67]: @ivanortegaalba it's a todo / temporary thing, the current VizPanel does not have a numerical id, we might need one for backward compatibility with the old dashboards but not added yet.

File: public/app/features/scenes/components/VizPanel/VizPanel.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/features/scenes/components/VizPanel/VizPanel.tsx:106]: @ivanortegaalba it does not change in this callback

File: public/app/features/scenes/querying/SceneQueryRunner.ts [hunk 6]
```diff
@@ -51,17 +57,52 @@ export class SceneQueryRunner extends SceneObjectBase<QueryRunnerState> {
...
-      this.querySub.unsubscribe();
-      this.querySub = undefined;
+    if (this._querySub) {
+      this._querySub.unsubscribe();
+      this._querySub = undefined;
+    }
+  }
+
+  public setContainerWidth(width: number) {
+    if (!width || !this.state.maxDataPointsFromWidth || this.state.maxDataPoints) {
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/scenes/querying/SceneQueryRunner.ts:90]: yea, agree. Updated this

**Comment by torkelo** [comment]
> Awesome work! Regarding migrations, have you thought anything about this long term? I guess we will use the schema pipeline to do the migration before returning the options to the client?

@mckn This PR is adding the same migrations that we have in the old PanelModel (ie calling the PanelPlugin onPanelMigration handler)

File: public/app/features/scenes/querying/SceneQueryRunner.test.ts (hunk 5 above)
Comment (torkelo) [reply public/app/features/scenes/querying/SceneQueryRunner.test.ts:74]: @ivanortegaalba Undefined, added another expect to make this more explicit (To make sure this call does not start queries right away )

**Pull request merged**